src/Libraries/Microsoft.Extensions.AI.OpenAI/OpenAIModelMapper.ChatMessage.cs

```diff
@@ -6,13 +6,97 @@
 
 using System;
 using System.Collections.Generic;
+using System.Linq;
 using System.Text.Json;
 using OpenAI.Chat;
 
 namespace Microsoft.Extensions.AI;
 
 internal static partial class OpenAIModelMappers
 {
+    internal static IEnumerable<ChatMessage> FromOpenAIChatMessages(IEnumerable<OpenAI.Chat.ChatMessage> inputs, JsonSerializerOptions options)
+    {
+        // Maps all of the OpenAI types to the corresponding M.E.AI types.
+        // Unrecognized or non-processable content is ignored.
+
+        foreach (OpenAI.Chat.ChatMessage input in inputs)
+        {
+            switch (input)
+            {
+                case SystemChatMessage systemMessage:
+                    yield return new ChatMessage
+                    {
+                        Role = ChatRole.System,
+                        AuthorName = systemMessage.ParticipantName,
+                        Contents = FromOpenAIChatContent(systemMessage.Content),
+                    };
+                    break;
+
+                case UserChatMessage userMessage:
+                    yield return new ChatMessage
+                    {
+                        Role = ChatRole.User,
+                        AuthorName = userMessage.ParticipantName,
+                        Contents = FromOpenAIChatContent(userMessage.Content),
+                    };
+                    break;
+
+                case ToolChatMessage toolMessage:
+                    string textContent = string.Join(string.Empty, toolMessage.Content.Where(part => part.Kind is ChatMessageContentPartKind.Text).Select(part => part.Text));
+                    object? result = textContent;
+                    if (!string.IsNullOrEmpty(textContent))
+                    {
+#pragma warning disable CA1031 // Do not catch general exception types
+                        try
+                        {
+                            result = JsonSerializer.Deserialize(textContent, options.GetTypeInfo(typeof(object)));
+                        }
+                        catch
+                        {
+                            // If the content can't be deserialized, leave it as a string.
+                        }
+#pragma warning restore CA1031 // Do not catch general exception types
+                    }
+
+                    yield return new ChatMessage
+                    {
+                        Role = ChatRole.Tool,
+                        Contents = new AIContent[] { new FunctionResultContent(toolMessage.ToolCallId, name: string.Empty, result) },
```

Comment:
Could we look up the name by matching the `ToolCallId` in any preceding `FunctionCallContent`?

I'm not sure what anyone will do with this name, just suggesting this to more closely align with the behavior we have in the opposite direction.

Reply:
I continue to question why FunctionResultContent needs the name at all.

Reply:
It became necessary at some point in one of the intermediate refactorings of `AIFunctionFactory` where parameter marshalling was the responsibility of the leaf clients. Given that we've pivoted back to encapsulating this inside the function itself, it should be possible to mark it optional again.

That being said, trying to recover the function name from the chat history seems reasonable to me.